plasmapy_nei/eigen/eigenclass.py

```diff
@@ -0,0 +1,360 @@
+"""
+Contains the class to provide access to eigenvalue data for the
+ionization and recombination rates.
+"""
+
+__all__ = ["EigenData"]
+
+import h5py
+import numpy as np
+from numpy import linalg as LA
+import pkg_resources
+import warnings
+
+try:
+    from plasmapy.particles import Particle, particle_input
+except (ImportError, ModuleNotFoundError):
+    from plasmapy.atomic import Particle, particle_input
+
+max_atomic_number = 30  # TODO: double check if this is the correct number
+
+
+def _get_equilibrium_charge_states(ioniz_rate, recomb_rate, natom):
+    """
+    Compute the equilibrium charge state distribution for the
+    temperature specified using ionization and recombination rates.
+
+    Parameters
+    ----------
+    ioniz_rate
+
+    recomb_rate
+
+    natom
+        The atomic number.
+    """
+    nstates = natom + 1
+    conce = np.zeros(nstates)
+    f = np.zeros(nstates + 1)
+    c = np.zeros(nstates + 1)
+    r = np.zeros(nstates + 1)
+
+    # The start index is 1.
+    for i in range(nstates):
+        c[i + 1] = ioniz_rate[i]
+        r[i + 1] = recomb_rate[i]
+
+    f[1] = 1.0
+    f[2] = c[1] * f[1] / r[2]
+
+    # simplest for hydrogen
+    if natom == 1:
+        f[1] = 1.0 / (1.0 + c[1] / r[2])
+        f[2] = c[1] * f[1] / r[2]
+        conce[0:2] = f[1:3]
+        return conce
+
+    # for other elements
+
+    for k in range(2, natom):
+        f[k + 1] = (-c[k - 1] * f[k - 1] + (c[k] + r[k]) * f[k]) / r[k + 1]
+
+    f[natom + 1] = c[natom] * f[natom] / r[natom + 1]
+
+    f[1] = 1.0 / np.sum(f)
+
+    f[2] = c[1] * f[1] / r[2]
+
+    for k in range(2, natom):
+        f[k + 1] = (-c[k - 1] * f[k - 1] + (c[k] + r[k]) * f[k]) / r[k + 1]
+
+    f[natom + 1] = c[natom] * f[natom] / r[natom + 1]
```

Comment:
```suggestion
    # f[0] = 0.0 from initialization
    f[1] = 1.0
    for k in range(2, natom+1):
        # note: this loop does NOT run for the hydrogen case, thus it does note defined f[2]
        f[k] = (-c[k - 2] / r[k]) * f[k - 2] \
               + ((c[k-1] + r[k-1]) / r[k]) * f[k-1])

    f[natom + 1] = c[natom] * f[natom] / r[natom + 1]
    
    # normalize the distribution
    f = f / np.sum(f)
```

There's a lot of unnecessary extra calculations being done here.  This should duplicate the previous code in a more concise and readable manner.